### PR TITLE
Added acr integration

### PR DESCRIPTION
**Pull request by aman-harness**

### CODE REVIEW

File: cmd/kaniko-acr/main.go [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+		fmt.Printf("acr token " + token)
+		err = createDockerCfgFile(username, token, registry)
+	}
+
+	return dockerConfig, nil
+}
+
+func getAcrToken(tenantId, clientId, clientSecret, certPath, registry string) (string, error) {
+	if tenantId == "" {
+		return "", fmt.Errorf("tenantId can't be empty foe AAD authentication")
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:312]: foe -> for

File: cmd/kaniko-acr/main.go [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+	"github.com/urfave/cli"
+
+	kaniko "github.com/drone/drone-kaniko"
+	"github.com/drone/drone-kaniko/pkg/artifact"
+	"github.com/drone/drone-kaniko/pkg/docker"
+)
+
+const (
+	dockerPath        string = "/kaniko/.docker"
+	accessKeyEnv      string = "AZURE_CLIENT_ID"
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:26]: rename variables

File: cmd/kaniko-acr/main.go [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+	secretKeyEnv      string = "AZURE_CLIENT_SECRET"
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:27]: ditto

File: docker/acr/Dockerfile.linux.amd64 [hunk 4]
```diff
@@ -0,0 +1,5 @@
+FROM gcr.io/kaniko-project/executor:v1.6.0
```
Comment (shubham149) [inline docker/acr/Dockerfile.linux.amd64:1]: use 1.8.1 version here

File: cmd/kaniko-acr/main.go [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch acr token")
+	}
+	var response map[string]interface{}
+	json.NewDecoder(jsonResponse.Body).Decode(&response)
+	return response["refresh_token"].(string), nil
+}
+
+// Create the docker config file for authentication
+func createDockerCfgFile(username, password, registry string) error {
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:366]: move this to common code

File: cmd/kaniko-acr/main.go [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+		"service":      {registry},
+		"tenant":       {tenantId},
+		"access_token": {token},
+	}
+	jsonResponse, err := http.PostForm(fmt.Sprintf("https://%s/oauth2/exchange", registry), formData)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch acr token")
+	}
+	var response map[string]interface{}
+	json.NewDecoder(jsonResponse.Body).Decode(&response)
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:361]: does it return err while decoding. If yes, handle it

File: cmd/kaniko-acr/main.go [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+	return response["refresh_token"].(string), nil
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:362]: ditto

File: cmd/kaniko-acr/main.go [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+
+	err := os.MkdirAll(dockerPath, 0600)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to create %s directory", dockerPath))
+	}
+
+	authBytes := []byte(fmt.Sprintf("%s:%s", username, password))
+	encodedString := base64.StdEncoding.EncodeToString(authBytes)
+	jsonBytes := []byte(fmt.Sprintf(`{"auths": {"%s": {"auth": "%s"}}}`, "https://"+registry, encodedString))
+	fmt.Print("-- " + username + " -- " + string(jsonBytes))
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:382]: remove it

File: cmd/kaniko-acr/main.go [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+	app.Action = run
+	app.Version = pluginVersion
+	app.Flags = []cli.Flag{
+		cli.StringFlag{
+			Name:   "dockerfile",
+			Usage:  "build dockerfile",
+			Value:  "Dockerfile",
+			EnvVar: "PLUGIN_DOCKERFILE",
+		},
+		cli.StringFlag{
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:53]: remove any flag that is not used

File: cmd/kaniko-acr/main.go [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+			Name:   "docker-username",
+			Usage:  "docker username",
+			EnvVar: "PLUGIN_USERNAME,DOCKER_USERNAME",
+		},
+		cli.StringFlag{
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:58]: ditto

File: cmd/kaniko-acr/main.go [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+	if tenantId == "" {
+		return "", fmt.Errorf("tenantId can't be empty foe AAD authentication")
+	}
+
+	if clientId == "" {
+		return "", fmt.Errorf("clientId can't be empty foe AAD authentication")
+	}
+
+	if clientSecret == "" && certPath == "" {
+		return "", fmt.Errorf("one of accessKey or secretKey should be defined")
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:320]: update the error messgae

File: cmd/kaniko-acr/main.go [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+	return dockerConfig, nil
+}
+
+func getAcrToken(tenantId, clientId, clientSecret, certPath, registry string) (string, error) {
+	if tenantId == "" {
+		return "", fmt.Errorf("tenantId can't be empty foe AAD authentication")
+	}
+
+	if clientId == "" {
+		return "", fmt.Errorf("clientId can't be empty foe AAD authentication")
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:316]: ditto

File: cmd/kaniko-acr/main.go [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+
+	if !noPush || clientId != "" {
+		// only setup auth when pushing or credentials are defined
+		if registry == "" {
+			return nil, fmt.Errorf("registry must be specified")
+		}
+
+		token, err := getAcrToken(tenantId, clientId, clientSecret, certPath, registry)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to fetch acrToken")
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:301]: acrToken. -> ACR token

File: cmd/kaniko-acr/main.go [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+	err := os.MkdirAll(dockerPath, 0600)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to create %s directory", dockerPath))
+	}
+
+	authBytes := []byte(fmt.Sprintf("%s:%s", username, password))
+	encodedString := base64.StdEncoding.EncodeToString(authBytes)
+	jsonBytes := []byte(fmt.Sprintf(`{"auths": {"%s": {"auth": "%s"}}}`, "https://"+registry, encodedString))
+	fmt.Print("-- " + username + " -- " + string(jsonBytes))
+	err = ioutil.WriteFile(dockerConfigPath, jsonBytes, 0644)
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:383]: We are writing to the same file at L249 as well

File: cmd/kaniko-acr/main.go [hunk 15]
```diff
@@ -0,0 +1,388 @@
...
+			Usage:  "Azure Tenant Id",
+			EnvVar: "TENANT_ID",
+		},
+		cli.StringFlag{
+			Name:   "client-id",
+			Usage:  "Azure Client Id",
+			EnvVar: "CLIENT_ID",
+		},
+		cli.StringFlag{
+			Name:   "assume-role",
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:157]: remove it

File: cmd/kaniko-acr/main.go [hunk 16]
```diff
@@ -0,0 +1,388 @@
...
+			Usage:  "Assume a role",
+			EnvVar: "PLUGIN_ASSUME_ROLE",
+		},
+		cli.StringFlag{
+			Name:   "external-id",
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:162]: ditto

File: cmd/kaniko-acr/main.go [hunk 17]
```diff
@@ -0,0 +1,388 @@
...
+	if clientSecret == "" && certPath == "" {
+		return "", fmt.Errorf("one of accessKey or secretKey should be defined")
+	}
+
+	// TODO check for presence of file as well.
+	os.Setenv(accessKeyEnv, clientId)
+	os.Setenv(secretKeyEnv, clientSecret)
+	os.Setenv(tenantKeyEnv, tenantId)
+	env, err := azidentity.NewEnvironmentCredential(nil)
+	context.Background()
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:328]: Remove it

File: cmd/kaniko-acr/main.go [hunk 18]
```diff
@@ -0,0 +1,388 @@
...
+
+	if clientSecret == "" && certPath == "" {
+		return "", fmt.Errorf("one of accessKey or secretKey should be defined")
+	}
+
+	// TODO check for presence of file as well.
+	os.Setenv(accessKeyEnv, clientId)
+	os.Setenv(secretKeyEnv, clientSecret)
+	os.Setenv(tenantKeyEnv, tenantId)
+	env, err := azidentity.NewEnvironmentCredential(nil)
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:327]: Handle err

File: cmd/kaniko-acr/main.go [hunk 19]
```diff
@@ -0,0 +1,388 @@
...
+		Scopes: []string{"https://management.azure.com/.default"},
+	}
+
+	os.Unsetenv(accessKeyEnv)
+	os.Unsetenv(secretKeyEnv)
+	os.Unsetenv(tenantKeyEnv)
+
+	azToken, err := env.GetToken(context.Background(), policy)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch access token")
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:339]: fix the error message

File: cmd/kaniko-acr/main.go [hunk 20]
```diff
@@ -0,0 +1,388 @@
...
+	}
+
+	acrToken, err := fetchAcrToken(tenantId, azToken.Token, registry)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch acr token")
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:344]: acr -> ACR

File: cmd/kaniko-acr/main.go [hunk 21]
```diff
@@ -0,0 +1,388 @@
...
+func fetchAcrToken(tenantId, token, registry string) (string, error) {
+	formData := url.Values{
+		"grant_type":   {"access_token"},
+		"service":      {registry},
+		"tenant":       {tenantId},
+		"access_token": {token},
+	}
+	jsonResponse, err := http.PostForm(fmt.Sprintf("https://%s/oauth2/exchange", registry), formData)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch acr token")
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:358]: ditto

File: cmd/kaniko-acr/main.go (hunk 1 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:323]: done

File: cmd/kaniko-acr/main.go (hunk 2 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:26]: done

File: cmd/kaniko-acr/main.go (hunk 3 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:27]: done

File: cmd/kaniko-acr/main.go (hunk 6 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:381]: done

File: cmd/kaniko-acr/main.go (hunk 7 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:382]: done

File: cmd/kaniko-acr/main.go (hunk 9 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:55]: done

File: cmd/kaniko-acr/main.go (hunk 10 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:60]: done

File: cmd/kaniko-acr/main.go (hunk 11 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:331]: done

File: cmd/kaniko-acr/main.go (hunk 12 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:327]: done

File: cmd/kaniko-acr/main.go (hunk 14 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:402]: removed

File: cmd/kaniko-acr/main.go (hunk 15 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:159]: removed

File: cmd/kaniko-acr/main.go (hunk 16 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:164]: removed

File: cmd/kaniko-acr/main.go (hunk 17 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:347]: removed

File: cmd/kaniko-acr/main.go (hunk 20 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:364]: done

File: cmd/kaniko-acr/main.go (hunk 21 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:378]: done

File: docker/acr/Dockerfile.linux.amd64 (hunk 4 above)
Comment (aman-harness) [reply docker/acr/Dockerfile.linux.amd64]: done

File: cmd/kaniko-acr/main.go (hunk 5 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go]: done

File: cmd/kaniko-acr/main.go (hunk 8 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go]: done

File: cmd/kaniko-acr/main.go (hunk 13 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go]: done

File: cmd/kaniko-acr/main.go (hunk 19 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:329]: this is correct. its called access token in azure

File: cmd/kaniko-acr/main.go [hunk 22]
```diff
@@ -0,0 +1,374 @@
...
+	if tenantId == "" {
+		return "", fmt.Errorf("tenantId can't be empty for AAD authentication")
+	}
+
+	if clientId == "" {
+		return "", fmt.Errorf("clientId can't be empty for AAD authentication")
+	}
+
+	if clientSecret == "" && cert == "" {
+		return "", fmt.Errorf("one of clientSecret or cert should be defined")
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:299]: clientSecret -> client secret

File: cmd/kaniko-acr/main.go [hunk 23]
```diff
@@ -0,0 +1,374 @@
...
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch ACR token")
+	}
+	var response map[string]interface{}
+	err = json.NewDecoder(jsonResponse.Body).Decode(&response)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to decode oauth exchange response")
+	}
+
+	fmt.Printf("refrsh token " + response["refresh_token"].(string))
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:356]: Remove this line

File: cmd/kaniko-acr/main.go [hunk 24]
```diff
@@ -0,0 +1,374 @@
...
+	if x, found := response["refresh_token"]; found {
+		return x.(string), nil
```
Comment (shubham149) [inline cmd/kaniko-acr/main.go:358]: also check whether x.(string) is valid or not. x is an interface here. So, we cannot assume that it is always a string.

File: pkg/docker/docker_file.go [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+	}
+
+	authBytes := []byte(fmt.Sprintf("%s:%s", username, password))
+	encodedString := base64.StdEncoding.EncodeToString(authBytes)
+	jsonBytes := []byte(fmt.Sprintf(`{"auths": {"%s": {"auth": "%s"}}}`, "https://"+registry, encodedString))
+	err = ioutil.WriteFile(path, jsonBytes, 0644)
+	if err != nil {
+		return errors.Wrap(err, "failed to create docker config file")
+	}
+	fmt.Print("crated docker file at " + path)
```
[the remark this answers is not in the document]
Comment (aman-harness) [reply pkg/docker/docker_file.go:33]: done

File: cmd/kaniko-acr/main.go (hunk 22 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go:299]: done

File: cmd/kaniko-acr/main.go (hunk 24 above)
Comment (aman-harness) [reply cmd/kaniko-acr/main.go]: done

File: docker/acr/Dockerfile.linux.amd64.kaniko1.8.1 [hunk 26]
```diff
@@ -0,0 +1,5 @@
+FROM gcr.io/kaniko-project/executor:v1.8.1
```
Comment (shubham149) [inline docker/acr/Dockerfile.linux.amd64.kaniko1.8.1:1]: remove this. 1.8.1 can be the default

File: docker/acr/manifest-kaniko1.8.1.tmpl [hunk 27]
```diff
@@ -0,0 +1,13 @@
+image: plugins/kaniko:{{#if build.tag}}{{trimPrefix "v" build.tag}}-kaniko1.8.1{{else}}latest-kaniko1.8.1{{/if}}
```
Comment (shubham149) [inline docker/acr/manifest-kaniko1.8.1.tmpl:1]: remove this

File: .drone.yml [hunk 28]
```diff
@@ -94,6 +111,23 @@ steps:
       exclude:
       - pull_request
 
+- name: acr-kaniko-v1-8
```
Comment (shubham149) [inline .drone.yml:114]: remove this

File: .drone.yml [hunk 29]
```diff
@@ -198,6 +244,18 @@ steps:
     username:
       from_secret: docker_username
 
+- name: manifest-acr
```
Comment (shubham149) [inline .drone.yml:247]: remove this.

File: docker/acr/Dockerfile.linux.amd64.kaniko1.8.1 (hunk 26 above)
Comment (aman-harness) [reply docker/acr/Dockerfile.linux.amd64.kaniko1.8.1:1]: done

File: docker/acr/manifest-kaniko1.8.1.tmpl (hunk 27 above)
Comment (aman-harness) [reply docker/acr/manifest-kaniko1.8.1.tmpl:1]: done

File: .drone.yml (hunk 29 above)
Comment (aman-harness) [reply .drone.yml:247]: done

File: .drone.yml (hunk 28 above)
Comment (shubham149) [reply .drone.yml:114]: The build will fail otherwise

File: docker/acr/manifest-kaniko1.8.1.tmpl [hunk 30]
```diff
@@ -0,0 +1,12 @@
+{{#if build.tags}}
```
Comment (shubham149) [inline docker/acr/manifest-kaniko1.8.1.tmpl:1]: remove this file

Comment (aman-harness) [reply docker/acr/manifest-kaniko1.8.1.tmpl]: done.

**Review by shubham149 — APPROVED** [review]

**Pull request merged**